lib/reference_model/README.md

```diff
@@ -0,0 +1,20 @@
+# Reference Model
+
+The reference model is built as a shell around Spike, with the goal of properly simulating asynchronous events.
+
+Spike is responsible for the functional execution, while the pipeline shell is responsible for timing interrupts and other asynchronous events, and timing the RVFI output for comparison with the core.
+
+TODO: Add a more detailed explenation
+
+## Usage 
+
+To run the reference model alongside cv32e40s, run `make` as normal, but include the following:
+
+- set `USE_RM=YES` to enable the reference model.
+
+- set `SPIKE_HOME=<path to riscv/riscv-isa-sim>` to use another spike version.
+If not set, spike from `vendor/riscv/riscv-isa-sim` will be used.
+
+For example:
+`make test TEST=hello-world USE_RM=YES SPIKE_HOME=~/core-v-verif/vendor/riscv/riscv-isa-sim`
```

Comment:
When running `make test TEST=hello-world USE_RM=1 SIMULATOR=vsim` I get
`# ** Error: (vsim-3763) SystemVerilog DPI cannot access file '/work/ropeders/src/core-v-verif-tne/tools/spike/lib/libriscv.so'`.

Spike was compiling for quite a time, but it seems it didn't succeed.
(If it fails at compiling, then I don't know why that isn't flagged to make, but that might be a problem with the "vendorization" and not with this PR.)
Here are a few messages that seemed relevant and might shine light on the situation:
```
make[1]: Leaving directory `/work/ropeders/src/core-v-verif-tne/vendor/riscv/riscv-isa-sim/build'
Makefile:349: warning: overriding recipe for target `disasm.o'
Makefile:349: warning: ignoring old recipe for target `disasm.o'
make[1]: Entering directory `/work/ropeders/src/core-v-verif-tne/vendor/riscv/riscv-isa-sim/build'
[...]
make[1]: Circular libfesvr.so <- libfesvr.so dependency dropped.
make[1]: Circular libcustomext.so <- libcustomext.so dependency dropped.
make[1]: Circular libsoftfloat.so <- libsoftfloat.so dependency dropped.
../scripts/mk-install-dirs.sh /work/ropeders/src/core-v-verif-tne/tools/spike/lib
```

@torjeikenes is this expected to work on the branch as it is now, or does it require a merge from `cv32e40s/release` first?

Reply:
Just merged in the new changes from srelease, which changes spike compilation slightly. Is this still a problem?

Reply:
Thanks. I am having other technical difficulties on my end, so I will try it out again once that is resolved.

Reply:
I'm trying `make test TEST=hello-world SIMULATOR=vsim USE_ISS=YES ISS=RM`,
after having done `git clean -xdff .` at the top directory,
then I get to:
```
../fesvr/fesvr_dpi.cc:22:10: fatal error: svdpi.h: No such file or directory
   22 | #include <svdpi.h>
```

Reply:
Right. You have to do `export QUESTASIM_HOME=<path to questasim>`. I think svdpi.h is included into spike [here](https://github.com/openhwgroup/core-v-verif/blob/3206c283fc50d32c4691df50eb9554e2584e2c03/vendor/riscv/riscv-isa-sim/Makefile.in#L72)

Reply:
Hmm, this does not happen to me... Could it be something [here](https://github.com/openhwgroup/core-v-verif/blob/3206c283fc50d32c4691df50eb9554e2584e2c03/mk/uvmt/vsim.mk#L200) since I don't have access to IMPERAS_HOME, but you do?

Reply:
> Right. You have to do `export QUESTASIM_HOME=<path to questasim>`. I think svdpi.h is included into spike [here](https://github.com/openhwgroup/core-v-verif/blob/3206c283fc50d32c4691df50eb9554e2584e2c03/vendor/riscv/riscv-isa-sim/Makefile.in#L72)

This is something from upstream "master" in core-v-verif, right?

Because spike started compiling fine before, and I did not get that error.
Plus, `lib/dpi_dasm` uses that same `svdpi.h` file and `make dpi_dasm SIMULATOR=vsim` works fine.
It just seems like it shouldn't be a technical necessity.

--------

> IMPERAS_HOME

I unsat everything IMPERAS and it got me even further.
`version 'GLIBC_2.28' not found`

This might be from the Red Hat version I am using.
If one specifies `ISS=RM` I don't think this system should be sensitive to whether imperas variables are set in the environment.

Reply:
@torjeikenes do you know how I could test just the vanilla spike "vendorization" without the refmodel? E.g. some other project like cva6 or e20. Or, if you know some other way, I would like to run something with just the vendorizationa and not the refmodel to see if the reason I can't run this stems from the vendorization itself or from something particular to the refmodel.

Reply:
Making sure I'm using the latest questasim and gcc (2024.1_1, 11.4.0).
I'm getting all kinds of different error messages no matter what I try.

```
make test TEST=hello-world SIMULATOR=vsim USE_ISS=YES ISS=RM
[...]
make[1]: Entering directory '/work/ropeders/src/core-v-verif-tne/vendor/riscv/riscv-isa-sim/build'
../scripts/mk-install-dirs.sh /work/ropeders/src/core-v-verif-tne/tools/spike/include
make[1]: *** No rule to make target 'config.h', needed by 'install-config-hdrs'.  Stop.
```

Reply:
> do you know how I could test just the vanilla spike "vendorization" without the remodel? 
You could attempt to build spike using the normal [build steps](https://github.com/riscv-software-src/riscv-isa-sim?tab=readme-ov-file#build-steps) from /vendor/riscv/riscv-isa-sim/ in either master or s/release

These steps work for me in s/release:

`cd core-v-verif/vendor/riscv/riscv-isa-sim`
`mkdir build`
`cd build`
`../configure --prefix=<core-v-verif path>/tools/spike`
`make`
`make install`

Reply:
> These steps work for me in s/release

Do they work after a fresh clone or `git reset . ; git checkout . ; git clean -xdff .`?

I get:
```
../fesvr/elfloader.cc:3:10: fatal error: config.h: No such file or directory
 #include "config.h"
          ^~~~~~~~~~
compilation terminated.
```
But when I grep for `config.h` there is no readme mentioning that I need to create it myself or anything like that.

Reply:
It seems like `config.h` is generated when running `../configure`. Do you have a `config.h` file in ..../riscv-isa-sim/build/ after running configure?

Reply:
No.

`configure` says:
```
config.status: creating config.h
config.status: config.h is unchanged
```

But there is no such file:
```
$ pwd
/work/ropeders/src/core-v-verif-tne/vendor/riscv/riscv-isa-sim
$ find | grep config.h
./config.h.in
```

Reply:
Hmm, that is weird. Are you able to clone and build vanilla spike https://github.com/riscv-software-src/riscv-isa-sim?

Reply:
No.
Maybe someone (Mike or Mario maybe) who are already able to compile that could sanity check this PR branch afterwards.
In the meantime, I think it is best to fix everything else, and then come back to this in the end.
I need to at least be able to run hello world and pass `ci_check` with Xcelium without changing anything in my standard setup, and all other review conversations should be resolved.

Reply:
For the record: I created this ticket https://github.com/riscv-software-src/riscv-isa-sim/issues/1683

Reply:
Well, I am back to errors like `failed: /lib64/libstdc++.so.6: version 'GLIBCXX_3.4.20' not found`.
I was close to getting it to work yesterday, but I don't know which planets need to be in alignment for this to work, so I will focus on testing Xcelium with Imperas for now.

Reply:
`ci_check` passes for me.
So when all feedback conversations are resolved I think we will be good.